### PR TITLE
Simplify query status control

### PR DESCRIPTION
**Pull request by zegnus**
## Investigation done ##

I've noticed an excessive number of invocations to the creation of the `FileDownloadInfo` object.

This is the analysis:
- Creation of the `FileDownloadInfo` object have a class level `synchronise`, along with an extra database query for the headers
- `FileDownloadInfo` is fully created but only the values `status` and `control` are used, the rest of the values are completely useless for this trace
- The `status` and `control` are queried at the method `checkPausedOrCreated` and invokated at `DownloadThread.transferData`.

So everytime that there is a transfer data, we call checkPausedOrCreated that creates a full FileDownloadInfo object in order to just check the status and the control.

## Solution implemented ##

I've created a slimmed down version that only creates an object with the `status` and `control` fields, removing all the unnecessary code that is not involved in the `checkPausedOrCreated` task

**Time for downloading 100MB from the Demo application**

Before | Now
--- | ---
6m12s | 5m

### CODE REVIEW

**Comment by hal9002** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://ci.novoda.com/job/download-manager-prb/132/
Test PASSed.

**Comment by blundell** [comment]
general idea seems sound to me :+1: didn't delve to deep in the code

File: library/src/main/java/com/novoda/downloadmanager/lib/FileDownloadInfo.java [hunk 1]
```diff
@@ -454,4 +454,48 @@ private Long getLong(String column) {
...
+                this.downloadsUriProvider = downloadsUriProvider;
+            }
+
+            public ControlStatus newControlStatus(long id) {
+                String[] projection = {DownloadContract.Downloads.COLUMN_CONTROL, DownloadContract.Downloads.COLUMN_STATUS};
+                Uri uri = ContentUris.withAppendedId(downloadsUriProvider.getAllDownloadsUri(), id);
+                Cursor downloadsCursor = contentResolver.query(uri, projection, null, null, null);
+                try {
+                    downloadsCursor.moveToFirst();
+                    int control = downloadsCursor.getInt(0);
```
Comment (xrigau) [inline library/src/main/java/com/novoda/downloadmanager/lib/FileDownloadInfo.java]: a bit flaky using raw ints here IMO

Comment (zegnus) [inline library/src/main/java/com/novoda/downloadmanager/lib/FileDownloadInfo.java]: The reason is because we do the projection so we don't need to query again to the cursor where are the columns, we have just done it at the projection line. As well is saves execution time for querying the column int when we have already specify them in the projection.

Comment (amlcurran) [inline library/src/main/java/com/novoda/downloadmanager/lib/FileDownloadInfo.java]: why?

**Comment by amlcurran** [comment]
Does it add any more allocations?

**Comment by zegnus** [comment]
This reduces allocations :)

**Comment by amlcurran** [comment]
Excellent :D

**Comment by amlcurran** [comment]
Ah, so we only create a ControlStatus object rather than a whole new FileDownloadInfo? Nice.

Just make sure it works OK with pause & resume and then LGTM :+1:

File: library/src/main/java/com/novoda/downloadmanager/lib/FileDownloadInfo.java [hunk 2]
```diff
@@ -454,4 +454,48 @@ private Long getLong(String column) {
...
+
+        public boolean isPaused() {
+            return control == DownloadsControl.CONTROL_PAUSED;
+        }
+
+        public boolean isCanceled() {
+            return status == DownloadStatus.CANCELED;
+        }
+
+        static final class Reader {
```
Comment (ouchadam) [inline library/src/main/java/com/novoda/downloadmanager/lib/FileDownloadInfo.java]: would prefer if we didn't follow this pattern and had separate repositories

Comment (zegnus) [inline library/src/main/java/com/novoda/downloadmanager/lib/FileDownloadInfo.java]: could have a discussion and modify this pattern in the FileDownloadInfo as well, but I would prefer if we do it in another PR

Comment (ouchadam) [inline library/src/main/java/com/novoda/downloadmanager/lib/FileDownloadInfo.java]: sounds good :+1:

**Comment by zegnus** [comment]
@amlcurran @ouchadam @xrigau all comments amended?

File: library/src/main/java/com/novoda/downloadmanager/lib/FileDownloadInfo.java [hunk 3]
```diff
@@ -454,4 +454,48 @@ private Long getLong(String column) {
             return cursor.getLong(cursor.getColumnIndexOrThrow(column));
         }
     }
+
+    static final class ControlStatus {
```
Comment (ouchadam) [inline library/src/main/java/com/novoda/downloadmanager/lib/FileDownloadInfo.java]: I like it!

**Comment by ouchadam** [comment]
LGTM :100:
